src/workerd/io/worker.c++

```diff
@@ -1060,6 +1113,46 @@ Worker::Isolate::Isolate(kj::Own<ApiIsolate> apiIsolateParam,
       }
     }
   });
+
+  // The PromiseCrossContextCallback is used to allow cross-IoContext promise following.
+  // When the IoContext::Scope is entered, we set the "promise context tag" associated
+  // with the IoContext on the Isolate that is locked. Any Promise that is created within
+  // that scope will be tagged with the same promise context tag. When an attempt to
+  // follow a promise occurs (e.g. either using Promise.prototype.then() or await, etc)
+  // our patched v8 logic will check to see if the followed promise's tag matches the
+  // current Isolate tag. If they do not, then v8 will invoke this callback. The promise
+  // here is the promise that belongs to a different IoContext.
+  lock->v8Isolate->SetPromiseCrossContextCallback([](v8::Local<v8::Context> context,
+                                                     v8::Local<v8::Promise> promise,
+                                                     v8::Local<v8::Object> tag) ->
+                                                       v8::MaybeLocal<v8::Promise> {
+    if (!IoContext::hasCurrent()) return promise;
```

Comment:
Maybe we should return a rejected promise saying that cross-context awaits don't work outside of a request context?

Reply:
Although I suppose _technically_ it might be OK because if the listener isn't in a request context, then its continuation isn't going to access anything inside the request context, so it's OK to run it in the publisher's context. But it might still be confusing, and I don't know what the use case would be, so maybe better to stick with an exception for now.